packages/engine/errors/2327.md

```diff
@@ -0,0 +1,19 @@
+---
+original: "Property '{0}' is optional in type '{1}' but required in type '{2}'."
+excerpt: "Property '{0}' in type '{2}' cannot be 'undefined'."
```

Comment:
I think this error can also show when you have:

`a: string | undefined`

being compared with:

`b?: string`.

That means that the excerpt above is slightly incorrect.

Reply:
@mattpocock Hm, it's a different error what you're referring to:
```
Type 'string | undefined' is not assignable to type 'string'.
```
The error 2327 happens instead of the above when you have set `exactOptionalPropertyTypes` to true in compiler options:

```ts
interface Foo {
   value?: string
 }
declare function test(foo: Foo): void

test({value: undefined}) // <<< Error!!!
test({}) // <<< OK
```

It errors with `exactOptionalPropertyTypes: true` because `?` has a different meaning when that option is enabled, it means `value` can only be `string` or the property `value` must not exist in the object at all.

With `exactOptionalPropertyTypes: true`, you'd type it like this to achieve the behavior without this option enabled:

```ts
interface Foo {
   value?: string | undefined // <<< '?' but also ' | undefined'
 }
declare function test(foo: Foo): void

test({value: undefined}) // <<< OK!
test({}) // <<< OK
```

So to sum up, with this option `exactOptionalPropertyTypes: true` there is a new concept of "optional" and "required" properties. "optional" ones cannot be set "undefined" (unless explicitly set to `undefined` like in the last snippet)

Does this make sense? 😅 
Maybe I should also mention that the user is seeing this because they have `exactOptionalPropertyTypes: true`? 🤔 
After writing this comment, maybe I can do better at explaining this cases here 🤔 .. I'll revisit it later.